### ISSUE TITLE
Implement hole punching 

### ISSUE THREAD

**Issue by DAOCUONG** [CympleTech/chamomile#23]
I would like to help implement hole punching for understanding i think we need stun server to know peer public ip and port. Client will send udp packing to server with port and adding it public ip and address to DHT So that others can connect directly to it. 

1)The two peers both open a UDP socket bound to a random local port
The two peers both contact a server on the internet. This server responds and tells them what IP and port their packet was received from. Since the server is on the Internet the IP/port seen by the server is the external IP and port
2)The two peers exchange information about the external IPs and ports through some mechanism.
The two peers start trying to send UDP traffic to each other.
3)Hopefully the external IP and port used by the NAT for the traffic to the server are the same as the external IP and port used for the traffic sent towards the other peer. So once both peers have sent packets out through their NATs the incoming packets match the mappings and peer to peer communication is established.

Note:
This strategy will work reliablly with full cone, restricted cone or restricted port cone NATs. It will work most of the time with port-preservative symmetric NATs. It will not work with randomising symmetric NATs.

Sun can you provide some guidance for this ???

**Comment by sunhuachuang** [comment]
That's cool. And I have some suggestions:
1. In our network, there are nodes with the role of relay (with public IP), and they can do the stun server work.
2. Now, NAT is automatically performed in the node, so when a node is connected, the NAT address of the node will be stored, which will help, when two nodes want to hole punching. And the relay (stun) node will has two functions: one is to tell the connector their external IP, and the other is to help the connector actively send data to the target node.

**Comment by DAOCUONG** [comment]
![Chamomile Hole Punching Relay Node](https://user-images.githubusercontent.com/5289711/167233219-53a03860-637b-422d-b15c-ba9b62640c9c.png)

**Comment by DAOCUONG** [comment]
@sunhuachuang  do we have message for sharing DHT Tablet yet? Relay node should have DHT record for alice and bob can   alice ask relay node to share the DHT records?

**Comment by DAOCUONG** [comment]
![Chamomile Hole Punching Peer Node](https://user-images.githubusercontent.com/5289711/167233386-56346108-0a9b-4e56-960e-683414b9ea91.png)

**Comment by DAOCUONG** [comment]
@sunhuachuang

**Comment by sunhuachuang** [comment]
Yep, We had the DHT share functions, https://github.com/CympleTech/chamomile/blob/main/src/server.rs#L285 and https://github.com/CympleTech/chamomile/blob/main/src/session.rs#L847
